docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+kind: Integration
+metadata:
+  name: my-integration
+spec:
+  api:
+    name: my-api
+  integrationType: HTTP_PROXY
+  integrationURI: https://example.org
+  integrationMethod: GET
+  payloadFormatVersion: "1.0"
+```
+
+#### API Generation Changes
+
+During CRD generation, we will inspect if any reference field needs to be added in
+CRD spec.
+
+```go
+// model.go
+// GetCRDs returns a slice of `CRD` structs that describe the
+// top-level resources discovered by the code generator for an AWS service API
+func (m *Model) GetCRDs() ([]*CRD, error) {
+	if m.crds != nil {
+		return m.crds, nil
+	}
+	crds := []*CRD{}
+
+	opMap := m.SDKAPI.GetOperationMap(m.cfg)
+
+	createOps := (*opMap)[OpTypeCreate]
+	readOneOps := (*opMap)[OpTypeGet]
+	readManyOps := (*opMap)[OpTypeList]
+	updateOps := (*opMap)[OpTypeUpdate]
+	deleteOps := (*opMap)[OpTypeDelete]
+	getAttributesOps := (*opMap)[OpTypeGetAttributes]
+	setAttributesOps := (*opMap)[OpTypeSetAttributes]
+
+	for crdName, createOp := range createOps {
+		if m.cfg.IsIgnoredResource(crdName) {
+			continue
+		}
+		crdNames := names.New(crdName)
+...
+		// Now any additional Spec fields that are required from other API
+		// operations.
+		for targetFieldName, fieldConfig := range m.cfg.ResourceFields(crdName) {
+			if fieldConfig.IsReadOnly {
+				// It's a Status field...
+				continue
+			}
+
++			if fieldConfig.Reference != nil {
++				// Add the spec field here
++				referenceFieldNames := names.New(targetFieldName)
++				rf := NewReferenceField(crd, referenceFieldNames, fieldConfig)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	cfg *ackgenconfig.FieldConfig,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReference"
+	gtp := "*ackv1alpha1.AWSResourceReference"
+	if cfg.Reference.IsList {
+		gt = "[]" + gt
+		gtp = "[]" + gtp
+	}
+	return &Field{
+		CRD:               crd,
+		Names:             fieldNames,
+		Path:              fieldNames.Original,
+		ShapeRef:          nil,
+		GoType:            gt,
+		GoTypeElem:        "AWSResourceReference",
+		GoTypeWithPkgName: gtp,
+		FieldConfig:       cfg,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // DereferenceResource finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       DereferenceResource(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
```

Comment:
Name is a bit confusing; I thought this was removing the reference. Something clearer would be `ResolveReference` , `ResolveReferenceFields`